src/cpp/dds/publisher/PublisherImpl.cpp

```diff
@@ -0,0 +1,492 @@
+
+    //REGISTER THE WRITER
+    rtps_participant_->registerWriter(impl->writer_, topic_att, writer_qos);
+
+    {
+        std::lock_guard<std::mutex> lock(mtx_writers_);
+        writers_[topic_att.getTopicDataType().to_string()].push_back(impl);
+    }
+
+    return writer;
+}
+
+bool PublisherImpl::delete_datawriter(
+        DataWriter* writer)
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    auto vit = writers_.find(writer->get_topic().getTopicName().to_string());
+    if (vit != writers_.end())
+    {
+        auto dw_it = std::find(vit->second.begin(), vit->second.end(), writer->impl_);
+        if (dw_it != vit->second.end())
+        {
+            vit->second.erase(dw_it);
+            return true;
+        }
+    }
+    return false;
+}
+
+DataWriter* PublisherImpl::lookup_datawriter(
+    const std::string& topic_name) const
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    auto it = writers_.find(topic_name);
+    if (it != writers_.end() && it->second.size() > 0)
+    {
+        return it->second.front()->user_datawriter_;
+    }
+    return nullptr;
+}
+
+bool PublisherImpl::get_datawriters(
+        std::vector<DataWriter*>& writers) const
+{
+    std::lock_guard<std::mutex> lock(mtx_writers_);
+    for (auto vit : writers_)
+    {
+        for (DataWriterImpl* dw : vit.second)
+        {
+            writers.push_back(dw->user_datawriter_);
+        }
+    }
+    return true;
+}
+
+/* TODO
+bool PublisherImpl::suspend_publications()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::resume_publications()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::begin_coherent_changes()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+/* TODO
+bool PublisherImpl::end_coherent_changes()
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+
+bool PublisherImpl::set_default_datawriter_qos(
+        const fastrtps::WriterQos& qos)
+{
+    if (&qos == &DATAWRITER_QOS_DEFAULT)
+    {
+        fastrtps::WriterQos def_qos;
+        default_datawriter_qos_.setQos(def_qos, true);
+    }
+    else if (default_datawriter_qos_.canQosBeUpdated(qos) && qos.checkQos())
+    {
+        default_datawriter_qos_.setQos(qos, false);
+        return true;
+    }
+    return false;
+}
+
+const fastrtps::WriterQos& PublisherImpl::get_default_datawriter_qos() const
+{
+    return default_datawriter_qos_;
+}
+
+/* TODO
+bool PublisherImpl::copy_from_topic_qos(
+        fastrtps::WriterQos&,
+        const fastrtps::TopicAttributes&) const
+{
+    logError(PUBLISHER, "Operation not implemented");
+    return false;
+}
+*/
+
+bool PublisherImpl::wait_for_acknowledgments(
+        const Duration_t& max_wait)
+{
+    Duration_t current = max_wait;
+    Duration_t begin, end;
+    std::lock_guard<std::mutex> lock(mtx_writers_);
```

Comment:
Is this mutex be locked for a long period?

Reply:
It can be locked up to `max_wait`.